File: src/python_client/fcmeans_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2024 AI&RD Research Group, Department of Information Engineering, University of Pisa
# SPDX-License-Identifier: Apache-2.0

import math
import os
import sys
import json
import time
import psutil
import traceback
import pandas as pd
import numpy as np
import _pickle as cPickle
from term import Atom
import python_common.common_utils as fl_utils
from numba import jit
from sklearn.metrics import adjusted_rand_score as ARI
from python_common.fedlang_process import FedLangProcess, start_process
from python_common.common_utils import logger_info, logger_error

FROBENIUS_NORM = 'fro'

start_time_fl = None
process = psutil.Process()

@jit(nopython=True)
def distance_fn(u: np.ndarray, v: np.ndarray):
    return np.linalg.norm(u - v)


def load_experiment_info(num_clients, target_feature, dataset_name = None):
    base_path = os.path.join(os.environ.get("PROJECT_PATH"), 'datasets')
    if dataset_name is None:
        dataset_name = 'pendigits.csv'
    dataset_path = os.path.join(base_path, dataset_name)
    df = pd.read_csv(dataset_path)
    size = int(math.ceil(df.shape[0] / num_clients))
    logger_info(f'dataset chunk size = {size}')
    y_true = df[str(target_feature)]
    df_X = df.drop(str(target_feature), axis=1)
    random_samples = np.random.randint(df.shape[0], size=size)
    values = df_X.values[random_samples, :]
    y_true = y_true.values[random_samples]
    logger_info(f'dataset chunk size = {size}, X.shape = {df_X.shape}, y.shape = {y_true.shape}')
    return values, y_true, target_feature


class FCMeansClientProcess(FedLangProcess):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def init_client(self, experiment, json_str_config: str):
        try:
            experiment_config = json.loads(json_str_config)
            fl_utils.logger.info(f"experiment = {experiment}, experiment_config = {experiment_config}")
            self.factor_lambda = float(experiment_config.get("lambdaFactor"))
            self.num_clients = experiment_config.get("numClients")
            self.target_feature = experiment_config.get("targetFeature")
            dataset_name = experiment_config.get("datasetName")
            self.X, self.y_true, self.target_feature = load_experiment_info(self.num_clients, self.target_feature, dataset_name)
            self.rows, self.num_features = self.X.shape
            self.get_node().send_nowait(sender=self.pid_,
                                        receiver=(Atom(_erl_client_name), Atom(_erl_worker_mailbox)),
                                        message=(Atom('fl_client_ready'), self.pid_))
            logger_info(f"after sending  ({_erl_worker_mailbox}, {_erl_client_name} )! (fl_client_ready)")
        except:
            logger_error(traceback.format_exc())

    def process_client(self, experiment, round_number, centers_param):
        global start_time_fl
        start_time = time.time()

        data = None

        try:
            logger_info(f"start process_client, experiment = {experiment}, round = {round_number}")
            centers_list = cPickle.loads(bytes(centers_param))
            #fl_utils.logger.info(f'{type(centers_list)} centers_list = {centers_list}')
            centers = np.array(centers_list)
            num_clusters = len(centers)
            num_objects = len(self.X)
            factor_lambda = self.factor_lambda

            ws = [[0] * self.num_features for i in range(num_clusters)]
            u = [0] * num_clusters

            for i in range(num_objects):
                denom = 0
                numer = [0] * num_clusters
                x = self.X[i]

                for c in range(num_clusters):
                    vc = centers[c]
                    numer[c] = (distance_fn(x, vc)) ** ((2) / (factor_lambda - 1))
                    if numer[c] == 0:
                        numer[c] = np.finfo(np.float).eps
                    denom = denom + (1 / numer[c])

                for c in range(num_clusters):
                    u_c_i = (numer[c] * denom) ** -1
                    ws[c] = ws[c] + (u_c_i ** factor_lambda) * x
                    u[c] = u[c] + (u_c_i ** factor_lambda)
            ws = [item.tolist() for item in ws]
            data = (u, ws)

            u = [[0] * num_objects for i in range(num_clusters)]

            for i in range(num_objects):

                denom = 0
                numer = [0] * num_clusters
                x = self.X[i]

                for c in range(num_clusters):
                    vc = centers[c]

                    numer[c] = (distance_fn(x, vc)) ** ((2) / (factor_lambda - 1))
                    if numer[c] == 0:
                        numer[c] = np.finfo(np.float64).eps
                    denom = denom + (1 / numer[c])

                for c in range(num_clusters):
                    u_c_i = (numer[c] * denom) ** -1
                    u[c][i] = u_c_i

            u = np.asarray(u).T
            y_pred = np.argmax(u, 1)

            ari_client = ARI(self.y_true, y_pred)

            total_memory, used_memory, free_memory = map(int, os.popen('free -t -m').readlines()[-1].split()[1:])
            memory_usage_percentage = round((used_memory/total_memory) * 100, 2)
            cpu_usage_percentage = psutil.cpu_percent()
            metrics_message = {
                "timestamp": int(time.time()), "type": "worker_metrics",
                "round": round_number,
                "client_id": os.environ.get("FL_CLIENT_ID"),
                "hostMetrics": {
                                "cpuUsagePercentage": cpu_usage_percentage,
                                "memoryUsagePercentage": memory_usage_percentage
                },
                "modelMetrics": {"ARI": ari_client}
            }

            self.get_node().send_nowait(sender=self.pid_,
                                        receiver=(Atom(_erl_client_name), Atom(_erl_worker_mailbox)),
                                        message=(Atom('fl_py_result'), cPickle.dumps(data), json.dumps(metrics_message)))
            logger_info(f"end process_client, experiment = {experiment}, round = {round_number}")
        except:
            logger_error(traceback.format_exc())

    def destroy(self):
        FedLangProcess.exit(self, None)
        logger_info('DESTROYYY')
        sys.exit(0)


if __name__ == "__main__":
    _pyrlang_node_id = sys.argv[1]
    _erl_client_name = sys.argv[2]
    _erl_worker_mailbox = sys.argv[3]
    _erl_cookie = sys.argv[4]
    experiment_id = sys.argv[5]
    client_id = os.environ.get("FL_CLIENT_ID")
    logger_info(f'pyrlang_node_id = {_pyrlang_node_id}, '
                         f'erl_client_name = {_erl_client_name}, '
                         f'erl_worker_mailbox = {_erl_worker_mailbox}, '
                         f'erl_cookie = {_erl_cookie}, '
                         f'experiment_id = {experiment_id}, '
                         f'client_id = {client_id}')
    start_process(FCMeansClientProcess, _pyrlang_node_id, _erl_cookie, _erl_client_name, _erl_worker_mailbox)

File: src/go/client/fcmeans_client.go
package main

import (
	"bytes"
	"encoding/csv"
	"encoding/gob"
	"encoding/json"
	"fcmeans/common"
	"log"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"github.com/ergo-services/ergo/etf"
	"github.com/shirou/gopsutil/cpu"
	"github.com/shirou/gopsutil/mem"
	"gonum.org/v1/gonum/mat"
)

var CLIENT_ID string

type FCMeansClient struct {
	factor_lambda   float64
	min_num_clients int
	target_feature  int
	X               [][]float64
	y_true          []float64
	rows            int
	num_features    int
	datasetName     string
	savedMetrics    []byte
	receivedData    [][]byte
	round_number    int
}
type FLExperiment struct {
	_global_model_parameters    [][]float64
	_client_configuration       map[string]interface{}
	_max_number_rounds          int
	_client_ids                 interface{}
	_type_of_termination        string
	_termination_threshold      float64
	_monitored_metric           interface{}
	_type_of_client_selection   interface{}
	_client_ratio               interface{}
	_client_values              interface{}
	_step_wise_client_selection bool
	_client_selection_threshold interface{}
	_round_selected_ids         []int
	_latency                    int
	_calls_list                 []etf.Tuple
	_validate                   bool
	experiment_history          interface{}
	_latency_required           bool
}
type ExperimentConfig struct {
	LambdaFactor  float64 `json:"lambdaFactor"`
	NumClients    int     `json:"numClients"`
	TargetFeature int     `json:"targetFeature"`
	DatasetName   string  `json:"datasetName"`
}

func distance_fn(A [][]float64) float64 {
	var sum float64
	for i := range A {
		for j := range A[i] {
			sum += A[i][j] * A[i][j]
		}
	}
	return math.Sqrt(sum)
}

func encodeToBytes(data interface{}) ([]byte, error) {
	var buf bytes.Buffer
	encoder := gob.NewEncoder(&buf)
	err := encoder.Encode(data)
	if err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func decodeFromBytes(data []byte, v interface{}) error {
	buf := bytes.NewBuffer(data)
	decoder := gob.NewDecoder(buf)
	return decoder.Decode(v)
}

func load_experiment_info(numClients int, targetFeature int, datasetName ...string) ([][]float64, []float64, int) {
	basePath := filepath.Join(os.Getenv("PROJECT_PATH"), "datasets")
	var datasetPath string
	//log.Printf("datasetName = %#v len = %d", datasetName, len(datasetName))
	if len(datasetName) <= 1 {
		datasetPath = filepath.Join(basePath, "pendigits.csv")
	} else {
		datasetPath = filepath.Join(basePath, datasetName[0])
	}
	log.Printf("dataset path = %s\n", datasetPath)
	file, err := os.Open(datasetPath)
	if err != nil {
		log.Fatalf("Failed to open dataset file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("Failed to read dataset file: %v", err)
	}

	size := int(math.Ceil(float64(len(records)-1) / float64(numClients)))
	log.Printf("dataset chunk size = %d\n", size)

	header := records[0]
	targetIndex := -1
	for i, col := range header {
		if col == strconv.Itoa(targetFeature) {
			targetIndex = i
			break
		}
	}
	if targetIndex == -1 {
		log.Fatalf("Target feature %s not found in dataset", targetFeature)
	}

	var dfX [][]float64
	var yTrue []float64
	for _, record := range records[1:] {
		var row []float64
		for i, value := range record {
			val, err := strconv.ParseFloat(value, 64)
			if err != nil {
				log.Fatalf("Failed to parse value: %v", err)
			}
			if i == targetIndex {
				yTrue = append(yTrue, val)
			} else {
				row = append(row, val)
			}
		}
		dfX = append(dfX, row)
	}

	randomSamples := rand.Perm(len(dfX))[:size]
	var values [][]float64
	var yTrueSample []float64
	for _, idx := range randomSamples {
		values = append(values, dfX[idx])
		yTrueSample = append(yTrueSample, yTrue[idx])
	}

	log.Printf("dataset chunk size = %d, X.shape = (%d, %d), y.shape = (%d)\n", size, len(dfX), len(dfX[0]), len(yTrueSample))
	return values, yTrueSample, targetFeature
}
func (f *FCMeansClient) Init_client(experiment string, json_str_config []byte, fp common.FedLangProcess) etf.Term {
	//log.Printf("experiment = %#v json_str_config = %#v\n", experiment, json_str_config)
	f.receivedData = make([][]byte, 0)
	var experimentConfig ExperimentConfig
	err := json.Unmarshal(json_str_config, &experimentConfig)
	if err != nil {
		log.Fatalf("Error parsing JSON config: %v", err)
	}

	log.Printf("experiment = %s, experimentConfig = %+v\n", experiment, experimentConfig)

	f.factor_lambda = experimentConfig.LambdaFactor
	f.min_num_clients = experimentConfig.NumClients
	f.target_feature = experimentConfig.TargetFeature
	f.datasetName = experimentConfig.DatasetName

	f.X, f.y_true, f.target_feature = load_experiment_info(f.min_num_clients, f.target_feature, f.datasetName)
	f.rows, f.num_features = len(f.X), len(f.X[0])
	/*
		f.Process.Send(
			gen.ProcessID{Name: f.Erl_worker_mailbox, Node: f.Erl_client_name},
			etf.Tuple{etf.Atom("fl_client_ready"), f.Process.Info().PID},
		)
	*/
	gob.Register([][]float64{})

	return etf.Tuple{etf.Atom("fl_client_ready"), fp.Own_pid}
}

func (f *FCMeansClient) Process_client(expertiment string, round_number int, centers_param []byte, fp common.FedLangProcess) etf.Term {

	log.Printf("start process_client, expertiment = %v, round_number = %v, centers_param_len = %v\n", expertiment, round_number, len(centers_param))
	var centers_list [][]float64
	// var decodedResult_tmp interface{}
	err := decodeFromBytes(centers_param, &centers_list)
	if err != nil {
		log.Println("Error decoding:", err)
		panic(err)
	}
	// for _, v := range decodedResult_tmp.([]interface{}) {
	// 	arr := make([]float64, 0)
	// 	for _, vv := range v.([]interface{}) {
	// 		arr = append(arr, vv.(float64))
	// 	}
	// 	centers_list = append(centers_list, arr)
	// }
	// log.Printf("centers_list = %v\n", centers_list)

	// Convert centers to a matrix
	centers := mat.NewDense(len(centers_list), len(centers_list[0]), nil)
	for i := 0; i < len(centers_list); i++ {
		centers.SetRow(i, centers_list[i])
	}

	// Initialize variables
	numClusters := centers.ColView(0).Len() // TODO: check if we numCluster is over rows instead of columns
	numObjects := len(f.X)                  //pls use more meaningful names
	factorLambda := f.factor_lambda
	numFeatures := f.num_features

	// Initialize ws and u slices
	ws := mat.NewDense(numClusters, numFeatures, nil)
	u := mat.NewVecDense(numClusters, nil)

	for i := 0; i < numObjects; i++ {
		denom := 0.0
		numer := make([]float64, numClusters)
		x := f.X[i]
		for j := 0; j < numClusters; j++ {
			vc := centers.RawRowView(j)
			numer[j] = math.Pow(distance_fn([][]float64{x, vc}), (2 / (factorLambda - 1)))
			if numer[j] == 0 {
				numer[j] = 1e-10 // TODO: this is not the correct translation check it
			}
			denom += (1 + numer[j])
		}
		for j := 0; j < numClusters; j++ {
			u_c_i := math.Pow((numer[j] * denom), -1)
			//ws[c] = ws[c] + (u_c_i ** factor_lambda) * x
			for k := 0; k < numFeatures; k++ {
				// the illness of this single line of code is beyond me
				// set the element at row j and column k to the sum of the
				// element at row j and column k and the product of the
				// element x[k] and the power of u_c_i to factorLambda
				// also does not return anything the panic is inside the function
				ws.Set(j, k, ws.At(j, k)+(math.Pow(u_c_i, factorLambda)*x[k]))
			}
			//u[c] = u[c] + (u_c_i ** factor_lambda)
			// Similar to the above line, set the element at index j to the
			// sum of the element at index j and the power of u_c_i to factorLambda
			u.SetVec(j, u.AtVec(j)+math.Pow(u_c_i, factorLambda))
		}
	}
	// Convert ws to a slice of slices and store data in a tuple
	wsSlice := make([][]float64, numClusters)
	for i := 0; i < numClusters; i++ {
		wsSlice[i] = ws.RawRowView(i)
	}
	data := []interface{}{
		u.RawVector().Data,
		wsSlice,
	}

	// Reinitialize u to a 2D slice with zeros
	uMatrix := make([][]float64, numClusters)
	for i := range uMatrix {
		uMatrix[i] = make([]float64, numObjects)
	}

	for i := 0; i < numObjects; i++ {
		denom := 0.0
		numer := make([]float64, numObjects)
		x := f.X[i]
		for j := 0; j < numClusters; j++ {
			vc := centers.RawRowView(j)
			numer[j] = math.Pow(distance_fn([][]float64{x, vc}), (2 / (factorLambda - 1)))
			if numer[j] == 0 {
				numer[j] = 1e-10 // TODO: his is not the correct translation check it
			}
			denom += (1 / numer[j])
		}
		for j := 0; j < numClusters; j++ {
			u_c_i := math.Pow((numer[j] * denom), -1)
			uMatrix[j][i] = u_c_i
		}
	}
	// u = np.asarray(u).T
	uMatrixT := make([][]float64, numObjects)
	for i := range uMatrixT {
		uMatrixT[i] = make([]float64, numClusters)
	}
	//y_pred = np.argmax(u, 1)
	y_pred := make([]float64, numObjects)
	for i := 0; i < numObjects; i++ {
		max := 0.0
		for j := 0; j < numClusters; j++ {
			if uMatrix[j][i] > max {
				max = uMatrix[j][i]
				y_pred[i] = float64(j)
			}
		}
	}
	// create a stub ARI score
	ari_client := 0
	v, _ := mem.VirtualMemory()
	total_memory := v.Total
	used_memory := v.Used
	memory_usage_percentage := math.Round((float64(used_memory)/float64(total_memory))*100) / 100

	cpu_percentages, _ := cpu.Percent(time.Second, false)

	metricsMessage := map[string]interface{}{
		"timestamp": time.Now().Unix(),
		"round":     round_number,
		"clientId":  fp.Own_pid,
		"hostMetrics": map[string]float64{
			"cpuUsagePercentage":    cpu_percentages[0],
			"memoryUsagePercentage": memory_usage_percentage,
		},
		"modelMetrics": map[string](float64){
			"ARI": float64(ari_client),
		},
	}
	data_bytes, err := encodeToBytes(data)
	if err != nil {
		panic(err)
	}
	metricsMessageBytes, _ := json.Marshal(metricsMessage)
	// s.Process.Send(
	// 	gen.ProcessID{Name: s.Erl_worker_mailbox, Node: s.Erl_client_name},
	// 	etf.Tuple{etf.Atom("process_server_ok"), buffer.Bytes(), metricsMessageBytes},
	// )
	log.Printf("end process_client, data_len = %v, metricsMessage = %v\n", len(data), metricsMessage)

	f.round_number = round_number
	fp.PeerSend(f.ringFunction(), "RingForward", data_bytes)
	// fp.PeerSend(f.ringFunction(), "RingForward", []byte("hello from "+CLIENT_ID))
	id, _ := strconv.Atoi(CLIENT_ID)
	if id != (round_number % 3) { // TODO: remove hardcoded number of clients
		return etf.Tuple{etf.Atom("fl_py_result_ack"), metricsMessageBytes}
	}
	f.savedMetrics = metricsMessageBytes
	return nil
}

func (f *FCMeansClient) ringFunction() func(int, int) int {
	rn := f.round_number
	return func(id, num_peers int) int {
		log.Printf("ringFunction: id = %d, num_peers = %d rn = %d\n", id, num_peers, rn)
		if id == (rn % num_peers) {
			log.Println("ringFunction: self send")
			return id
		}

		peer_id := (id + 1) % num_peers
		log.Printf("ringFunction: peer_id = %d\n", peer_id)
		return peer_id
	}
}

func (f *FCMeansClient) RingForward(data_bytes interface{}, fp common.FedLangProcess) {
	log.Printf("RingForward: received msg = %#v\n", data_bytes)
	id, _ := strconv.Atoi(CLIENT_ID)
	log.Printf("RingForward: id = %d, round_number = %d num_clients = %d\n", id, f.round_number, fp.NumClients)
	if id != (f.round_number % fp.NumClients) {
		log.Printf("RingForward: not my turn, forwarding\n")
		fp.PeerSend(f.ringFunction(), "RingForward", data_bytes)
		return
	}
	// master peer should aggregate the results
	// TODO: for now it just sends all the results once they all arrive
	f.receivedData = append(f.receivedData, data_bytes.([]byte))
	log.Printf("RingForward: received data len = %d\n", len(f.receivedData))
	if fp.NumClients == 0 || len(f.receivedData) < fp.NumClients {
		return
	}
	log.Printf("RingForward: all results have arrived\n")
	// all results have arrived
	for _, data := range f.receivedData {
		// process the data
		fp.WorkerSend(etf.Tuple{etf.Atom("fl_py_result"), data, f.savedMetrics})
		// break // for now, just process the first result TODO: aggregate the results
	}
	f.receivedData = make([][]byte, 0)
}

func (f *FCMeansClient) Destroy(fp common.FedLangProcess) {
	log.Printf("DESTROYYYY")
	os.Exit(0)
}

// func (f *FCMeansClient) Update_graph(clients etf.Term, fp common.FedLangProcess) {
// 	log.Printf("Update_graph")
// 	log.Printf("clients = %#v\n", clients)
// }

func main() {

	go_node_id := os.Args[1]         // go_c0ecdfb7-00f1-4270-8e46-d835bd00f153@127.0.0.1
	erl_client_name := os.Args[2]    // director@127.0.0.1
	erl_worker_mailbox := os.Args[3] // mboxserver_c0ecdfb7-00f1-4270-8e46-d835bd00f153
	erl_cookie := os.Args[4]         // cookie_123456789
	experiment_id := os.Args[5]      // c0ecdfb7-00f1-4270-8e46-d835bd00f153
	CLIENT_ID = os.Args[6]           // 0

	log.Printf("gorlang_node_id = %v, erl_client_name = %v, erl_worker_mailbox = %v, erl_cookie = %v, experiment_id = %v\n", go_node_id, erl_client_name, erl_worker_mailbox, erl_cookie, experiment_id)

	logFileName := os.Getenv("FL_CLIENT_LOG_FOLDER") + "/" + CLIENT_ID + ".log"
	if logFileName == "" {
		logFileName = "default_client.log"
	}

	logFile, err := os.OpenFile(logFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		log.Fatalf("Failed to open log file: %v", err)
	}
	defer logFile.Close()
	log.SetOutput(logFile)

	common.StartProcess[FCMeansClient](go_node_id, erl_cookie, erl_client_name, erl_worker_mailbox, experiment_id)
}
